### ISSUE TITLE
SEGFAULT when Ipopt interface is used from MATLAB R2015b

### ISSUE THREAD

**Issue by mkotlyar** [casadi/casadi#1762]
To reproduce:
```matlab
x = casadi.MX.sym('x');
y = (x - 3).^2;
nlp.x = x;
nlp.f = y;
solver = casadi.nlpsol('solver', 'ipopt', nlp);
solver()
```
Output:
[matlab_crash.txt](https://github.com/casadi/casadi/files/217096/matlab_crash.txt)

**Comment by mkotlyar** [comment]
The crash happens inside a BLAS routine in the MKL library distributed with MATLAB. Could it be the MKL version mismatch?
```
Stack Trace (from fault):
[  0] 0x00007fd04469190c        /usr/local/MATLAB/R2015b/bin/glnxa64/mkl.so+33696012 mkl_blas_cnr_def_xdcopy+00000252
[  1] 0x00007fd042adc678        /usr/local/MATLAB/R2015b/bin/glnxa64/mkl.so+04642424 mkl_blas_dcopy+00000200
[  2] 0x00007fd042942b4c        /usr/local/MATLAB/R2015b/bin/glnxa64/mkl.so+02964300 dcopy_+00000080
[  3] 0x00007fd051c8c841                       /usr/local/lib/libipopt.so.0+01890369 _ZN5Ipopt11IpBlasDcopyEiPKdiPdi+00000129
```

**Comment by jgillis** [comment]
Is this using a binary or source build?

**Comment by mkotlyar** [comment]
@jgillis casadi and Ipopt built from sources.

**Comment by jgillis** [comment]
If you compile blas/lapack stastically into ipopt, you don't have this problem : https://github.com/casadi/testbot/blob/master/recipes/ipopt.sh#L20

**Comment by mkotlyar** [comment]
Will try it. I guess it should be enough to specify `--with-blas=BUILD --with-lapack=BUILD` when `configure`ing Ipopt?

**Comment by jgillis** [comment]
I think you need `--disable-shared` as well

**Comment by mkotlyar** [comment]
@jgillis I rebuild Ipopt using the script you gave me (I only changed the version number and install prefix). I am still getting the same crash:
[Uploading matlab_crash_dump.27372-1.txt…]()

I checked which modules does the new libipopt depend on and got this:
```
ldd /usr/local/lib/libipopt.so.0
	linux-vdso.so.1 =>  (0x00007ffca4401000)
	libcoinmumps.so.0 => /usr/local/lib/libcoinmumps.so.0 (0x00007f7d0602f000)
	libcoinhsl.so.0 => /usr/local/lib/libcoinhsl.so.0 (0x00007f7d05d07000)
	liblapack.so.3 => /usr/lib/liblapack.so.3 (0x00007f7d05567000)
	libblas.so.3 => /usr/lib/libblas.so.3 (0x00007f7d05307000)
	libdl.so.2 => /lib/x86_64-linux-gnu/libdl.so.2 (0x00007f7d050ff000)
	libstdc++.so.6 => /usr/lib/x86_64-linux-gnu/libstdc++.so.6 (0x00007f7d04d77000)
	libm.so.6 => /lib/x86_64-linux-gnu/libm.so.6 (0x00007f7d04a6f000)
	libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f7d0469f000)
	/lib64/ld-linux-x86-64.so.2 (0x000055b0172f2000)
	libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f7d04487000)
	libcoinmetis.so.0 => /usr/local/lib/libcoinmetis.so.0 (0x00007f7d04237000)
	libgfortran.so.3 => /usr/lib/x86_64-linux-gnu/libgfortran.so.3 (0x00007f7d03f07000)
	libgomp.so.1 => /usr/lib/x86_64-linux-gnu/libgomp.so.1 (0x00007f7d03cdf000)
	libopenblas.so.0 => /usr/lib/libopenblas.so.0 (0x00007f7d01d1f000)
	libquadmath.so.0 => /usr/lib/x86_64-linux-gnu/libquadmath.so.0 (0x00007f7d01adf000)
	libpthread.so.0 => /lib/x86_64-linux-gnu/libpthread.so.0 (0x00007f7d018bf000)
```

**Comment by jgillis** [comment]
Does the binary work for this example?

**Comment by mkotlyar** [comment]
What do you mean by `does it work`? It prints this
```
This is Ipopt version 3.12, running with linear solver ma27.

Number of nonzeros in equality constraint Jacobian...:        0
Number of nonzeros in inequality constraint Jacobian.:        0
Number of nonzeros in Lagrangian Hessian.............:        1
```
and then crashes.

**Comment by mkotlyar** [comment]
I still don't get why does `libipopt` call something from `mkl.so`.

**Comment by jgillis** [comment]
Let me rephrase the question; could you check if the example your provided crashes also when you using the casadi binaries as opposed to casadi source build

**Comment by mkotlyar** [comment]
I tried this binary http://files.casadi.org/3.0.0/linux/casadi-matlabR2014b-v3.0.0.tar.gz, and it works.

**Comment by jgillis** [comment]
Have you tried compiling matlab with '-DWITH_DEEPBIND=ON'?

**Comment by mkotlyar** [comment]
I will try it. What I did now, I rebuilt CasADi after I have recompiled Ipopt, and now it crashes on
```matlab
solver = casadi.nlpsol('solver', 'ipopt', nlp);
```

[matlab_crash_dump.2504-1.txt](https://github.com/casadi/casadi/files/217343/matlab_crash_dump.2504-1.txt)

```
Stack Trace (from fault):
[  0] 0x00007f4decf80397                       /usr/local/lib/libipopt.so.0+00398231 _ZN5Ipopt16IpoptApplication14OpenOutputFileENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEENS_13EJournalLevelE+00000055
[  1] 0x00007f4ded3b2f97           /usr/local/lib/libcasadi_nlpsol_ipopt.so+00081815 _ZNK6casadi14IpoptInterface11init_memoryEPv+00002759
```

**Comment by jgillis** [comment]
Could it be that you compiled one with `gcc-4.*` and another with `gcc-5.*`?
There's really quite some stuff that can go wrong; that's why I created the binaries ;-)

**Comment by mkotlyar** [comment]
I compiled both today with `g++ (Ubuntu 5.2.1-22ubuntu2) 5.2.1 20151010`.

**Comment by jgillis** [comment]
Should you feel like more debugging, consider checking the differences between your build script and ours: https://github.com/casadi/testbot/blob/master/recipes/ipopt.sh, https://github.com/casadi/binaries/blob/debian-matlab/.travis.yml

**Comment by mkotlyar** [comment]
I rebuilt CasADi with `-DWITH_DEEPBIND=ON` and `casadi.nlpsol(...)` is crashing in the same way.

@jgillis do you have any idea why Ipopt would call something from `mkl.so`?

**Comment by mkotlyar** [comment]
btw, now it fails not only in MATLAB, but in Python as well.

**Comment by jaeandersson** [comment]
If you get the initial printout from IPOPT, it is a clear indication that the failure is related to the linear solver that IPOPT uses, here ma27. How did you install HSL?

**Comment by mkotlyar** [comment]
The following sequence of actions has fixed the issue for me:

(1) Configure, rebuild and install Ipopt from scratch:
```
tar -xvzf Ipopt-3.12.4.tgz
cd Ipopt-3.12.4
mkdir build
cd build
../configure --prefix=/usr/local
make -j 4
sudo make install
```
(2) Reconfigure CasADi with `-DWITH_DEEPBIND=ON`, build and install.

**Comment by ghorn** [comment]
What does DEEPBIND do?

**Comment by jaeandersson** [comment]
> What does DEEPBIND do?

http://linux.die.net/man/3/dlopen

> RTLD_DEEPBIND (since glibc 2.3.4)
> Place the lookup scope of the symbols in this library ahead of the global scope. This means that a self-contained library will use its own symbols in preference to global symbols with the same name contained in libraries that have already been loaded. This flag is not specified in POSIX.1-2001.
